### PR TITLE
incr. comp.: Don't export impl_stable_hash_via_hash!() and warn about using it.

### LINKED ISSUE #96013: Remove impl_stable_hash_via_hash!()
I came across `impl_stable_hash_via_hash!()` the other day and it seems very problematic because one can easily generate invalid `HashStable` implementations without even noticing:

```rust

#[derive(Hash)]
struct Foo<'tcx> {
    def_id: DefId, // numeric value of def_id ends up in fingerprint,
    ptr: *const c_void, // memory address ends up in fingerprint,
    bar: Interned<'tcx, Bar>, // ditto
}

// Yay, three separate bugs in a single line!
impl_stable_hash_via_hash!(Foo);
```

I think we should just remove the macro (or keep it local to the module that implements `HashStable` for basic types). `#[derive(HashStable_Generic)]` is available as a safe alternative.

`rustc_lint_defs::Level` looks like a case where this is already a problem because of the `HirId` in the contained `LintExpectationId`.

cc @rust-lang/wg-incr-comp

### PR DESCRIPTION
**Pull request by michaelwoerister**
Fixes https://github.com/rust-lang/rust/issues/96013.

### CODE REVIEW

File: compiler/rustc_errors/src/lib.rs [hunk 1]
```diff
@@ -1548,7 +1549,19 @@ pub fn add_elided_lifetime_in_path_suggestion(
 
 /// Useful type to use with `Result<>` indicate that an error has already
 /// been reported to the user, so no need to continue checking.
-#[derive(Clone, Copy, Debug, Encodable, Decodable, Hash, PartialEq, Eq, PartialOrd, Ord)]
+#[derive(
```
Comment (compiler-errors) [inline compiler/rustc_errors/src/lib.rs:1552]: nit: this probably should be split into two derives so it wraps more neatly

Comment (michaelwoerister) [reply compiler/rustc_errors/src/lib.rs]: Yes, that's better.

**Comment by michaelwoerister** [comment]
Removed the `#[inline]` annotations on the non-trivial `hash_stable()` methods. Let's see if that gets rid of the performance regressions.

@bors try @rust-timer queue

**Comment by michaelwoerister** [comment]
This looks better. The PR should be ready for review now.

**Comment by compiler-errors** [comment]
Looks good

@bors r+

**Comment by michaelwoerister** [comment]
Hm, that regression is unexpected and did not show up in previous perf run. Could it be spurious?

**Comment by michaelwoerister** [comment]
Yes, the `syn-1.0.89` benchmark looks flaky at the moment and the regression has gone away in subsequent perf runs.
@rustbot label: +perf-regression-triaged

**Comment by rylev** [comment]
Linking with https://github.com/rust-lang/rustc-perf/issues/1301 which tracks the issue of `syn-1.0.89 opt` being flaky.

**Pull request merged**